gadget/filesystemimage.go

```diff
@@ -0,0 +1,133 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+)
+
+type MkfsFunc func(imgFile, label, contentsRootDir string) error
+
+var (
+	mkfsHandlers = map[string]MkfsFunc{
+		"vfat": MkfsVfat,
+		"ext4": MkfsExt4,
+	}
+)
+
+// FilesystemImageWriter is capable of creating filesystem images described by
+// positioned structures.
+type FilesystemImageWriter struct {
+	contentDir string
+	ps         *PositionedStructure
+	workDir    string
+}
+
+// PostStageFunc is called after the filesystem contents for the given structure
+// have been staged at a temporary location
+type PostStageFunc func(rootDir string, ps *PositionedStructure) error
+
+// Write creates the filesystem inside the provided image file and populates it
+// with data according to content declartion of the structure. Content data is
+// staged in a temporary location. An optional post-stage helper function can be
+// used to manipulate the data before it is copied over to the image.
+func (f *FilesystemImageWriter) Write(fname string, postStage PostStageFunc) error {
+	st, err := os.Stat(fname)
+	if err != nil {
+		return fmt.Errorf("cannot stat image file: %v", err)
+	}
+	if sz := st.Size(); sz != int64(f.ps.Size) {
+		return fmt.Errorf("size of image file %v is different from declared structure size %v", sz, f.ps.Size)
+	}
+
+	mkfs := mkfsHandlers[f.ps.Filesystem]
+	if mkfs == nil {
```

Comment:
A remark/observation: In theory this is already guaranteed by `NewFilesystemImageWriter` at construction time, in practice  we're keeping PositionedStructure pointer and nothing prevents it from getting modified (and ending up with unsupported fs after ctor) since all members are public.

Also, afaict, this error case doesn't have a test atm.

Reply:
There's one here: https://github.com/snapcore/snapd/pull/7041/files#diff-399ad2ddfb72083e771e374d24de15e4R259

Reply:
Right, missed that one because of the different error message ;)